src/core/server/http/http_server.test.ts

```diff
@@ -1431,3 +1434,71 @@ describe('setup contract', () => {
     });
   });
 });
+
+describe('Graceful shutdown', () => {
+  test('it should wait until the ongoing requests are resolved and reject any new ones', async () => {
+    const gracefulShutdownTimeout = config.gracefulShutdownTimeout.asMilliseconds();
+
+    const { registerRouter, server: innerServer } = await server.setup(config);
+
+    const router = new Router('', logger, enhanceWithContext);
+    router.post(
+      {
+        path: '/',
+        validate: { body: schema.object({ test: schema.number() }) },
+        options: { body: { accepts: 'application/json' } },
+      },
+      async (context, req, res) => {
+        // It takes to resolve the same period of the gracefulShutdownTimeout.
+        // Since we'll trigger the stop a few ms after, it should have time to finish
+        await new Promise((resolve) => setTimeout(resolve, gracefulShutdownTimeout));
+        return res.ok({ body: req.route });
+      }
+    );
+    registerRouter(router);
+
+    await server.start();
+
+    const makeRequest = () => supertest(innerServer.listener).post('/').send({ test: 1 });
+
+    const [firstResponse, , secondResponse] = await Promise.all([
+      // Trigger a request that should succeed
+      makeRequest().expect(200, {
+        method: 'post',
+        path: '/',
+        options: {
+          authRequired: true,
+          xsrfRequired: true,
+          tags: [],
+          timeout: {
+            payload: 10000,
+          },
+          body: {
+            parse: true, // hapi populates the default
+            maxBytes: 1024, // hapi populates the default
+            accepts: ['application/json'],
+            output: 'data',
+          },
+        },
+      }),
+      // Stop the server while the request is in progress
+      (async () => {
+        await new Promise((resolve) => setTimeout(resolve, gracefulShutdownTimeout / 3));
```

Comment:
Shouldn't we test that the server responds to a connection created *before* the server shut down initiated?

Reply:
That's what I'm doing:
- `firstResponse` is an ongoing request created before the shutdown signal occurred, and it should be resolved with 200.
- `secondResponse` is a request coming in after the shutdown signal occurred. It should be rejected with 503.

I've split the tests to make it more clear.